[Docs][Kuberay] Documentation for Using Kuberay with TPUs (#45924)
ray-project/ray: doc/source/cluster/kubernetes/examples/tpu-serve-stable-diffusion.md

## Step 4: View the Serve deployment in the Ray dashboard

Verify the RayService is deployed and running

```sh
kubectl get rayservice

# NAME               SERVICE STATUS   NUM SERVE ENDPOINTS
# stable-diffusion-tpu-serve-svc   Running          2
```

Port-forward the Ray dashboard from the Ray head service. To view the dashboard, open http://localhost:8265/ on your local machine.

```sh
kubectl port-forward svc/stable-diffusion-tpu-head-svc 8265:8265 &
```

You may now monitor the status of the RayService in the Ray dashboard from the the 'Serve' tab. The installed RayService
should create a running application with the name 'stable_diffusion'. The Ray Serve application will have two Ray Serve deployments, the
API ingress which will receive input prompts and the Stable Diffusion model server.

> 

Reviewer: how about adding a screenshot of the dashboard here?

Author: Done in [f5d27f4](https://github.com/ray-project/ray/pull/45924/commits/f5d27f4112c93262a135572a45a36b30b365faa2), I also updated some of the instructions (there was a step missing to create a python virtual environment and install the needed dependencies to run the prompt script).

## Step 3: Install the RayService

```sh
# Creates a RayCluster with a single-host v4 TPU worker group of 2x2x1 topology
kubectl apply -f https://raw.githubusercontent.com/ray-project/kuberay/master/ray-operator/config/samples/ray-service.tpu-single-host.yaml
```

KubeRay operator v1.1.0 adds a new `NumOfHosts` field to the RayCluster CR, supporting multi-host worker groups. This field specifies the number of workers to create per replica, with each replica representing a multi-host PodSlice. The value for `NumOfHosts` should match the number of TPU VM hosts expected by the given `cloud.google.com/gke-tpu-topology` node selector. For this example, the Stable Diffusion model is small enough to run on a single TPU host, so `numOfHosts` is set to 1 in the RayService manifest.

## Step 4: View the Serve deployment in the Ray dashboard

Verify the RayService is deployed and running

```sh
kubectl get rayservice

# NAME               SERVICE STATUS   NUM SERVE ENDPOINTS
# stable-diffusion-tpu-serve-svc   Running          2
```

Retrieve the name of the RayCluster Head service

```sh

> kubectl get svc

Author: This command is no longer necessary with [3deb95f](https://github.com/ray-project/ray/pull/45924/commits/3deb95f8dc973c17cb8bfdd572a542ca5b986708).

```

Then, port-forward the Ray dashboard. To view the dashboard, open http://localhost:8265/ on your local machine.

```sh
# change the service name to match the one output under `kubectl get svc`

> kubectl port-forward svc/stable-diffusion-tpu-raycluster-ljw2r-head-svc 8265:8265 &

Author: Done in [3deb95f](https://github.com/ray-project/ray/pull/45924/commits/3deb95f8dc973c17cb8bfdd572a542ca5b986708)

## Step 1: Create a Kubernetes cluster with TPUs

Follow [Creating a GKE Cluster with TPUs for KubeRay](kuberay-gke-tpu-cluster-setup) to create a GKE Autopilot or Standard cluster with 1 CPU node and 2 TPU nodes.

## Step 2: Install the KubeRay operator

Follow [this document](kuberay-operator-deploy) to install the latest stable KubeRay operator via Helm repository. Multi-host TPU support is provided in KubeRay v1.1.0+. Please note that the YAML file in this example uses `serveConfigV2`, which is supported starting from KubeRay v0.6.0.

## Step 3: Create a RayCluster with a TPU worker group

```sh
# Creates a RayCluster with a single-host v4 TPU worker group of 2x2x1 topology
kubectl apply -f https://raw.githubusercontent.com/ray-project/kuberay/v1.1.1/ray-operator/config/samples/ray-cluster.tpu-v4-singlehost.yaml

# Creates a RayCluster with a multi-host v4 TPU worker group of 2x2x2 topology

> kubectl apply -f https://raw.githubusercontent.com/ray-project/kuberay/v1.1.1/ray-operator/config/samples/ray-cluster.tpu-v4-multihost.yaml

Author: I think it should behave the same, the multi-host setup just requires the webhook for JAX initialization and allows for distributing the workload across multiple VMs.

```

KubeRay operator v1.1.0 adds a new `NumOfHosts` field to the RayCluster CR, supporting multi-host worker groups. This field specifies the number of workers to create per replica, with each replica representing a multi-host PodSlice. The value for `NumOfHosts` should match the number of TPU VM hosts expected by the given `cloud.google.com/gke-tpu-topology` node selector.

## Step 4: Run the Serve deployment with ray job submit

Download the Ray Serve and API client Python files:

```sh
git clone https://github.com/ray-project/serve_config_examples.git

cd stable_diffusion/tpu
```

Retrieve the name of the RayCluster Head service

```sh
kubectl get svc
```

Then, port-forward the Ray dashboard. To view the dashboard, open http://localhost:8265/ on your local machine.

```sh
kubectl port-forward svc/example-cluster-kuberay-head-svc 8265:8265 &
```

In a separate terminal, submit the serve deployment using ray job.

```sh

> ray job submit --runtime-env runtime_env.yaml -- python3 ray_serve_diffusion_flax.py

Reviewer 2: The examples from https://github.com/ray-project/serve_config_examples usually bunded with a serve config in a RayService, is it possible to do the same thing instead of submitting `ray_serve_diffusion_flax.py` as a job?

Author: Sure that should be possible, I'll test it out and update the PR.

(kuberay-tpu-stable-diffusion-example)=

# Serve a Stable Diffusion model on GKE with TPUs

> > **Note:** The Python files for the Ray Serve app and its client are in the [ray-project/serve_config_examples](https://github.com/ray-project/serve_config_examples). This example is adapted from the [tensorflow/tpu](https://github.com/tensorflow/tpu/tree/master/tools/ray_tpu/src/serve) Cloud TPU example.

Reviewer 3: ```suggestion
> **Note:** The Python files for the Ray Serve app and its client are in the [ray-project/serve_config_examples](https://github.com/ray-project/serve_config_examples). This example adapts the [tensorflow/tpu](https://github.com/tensorflow/tpu/tree/master/tools/ray_tpu/src/serve) to use cloud TPUs.
```

Reviewer 3: Is this rewording still technically accurate?

Author: Both examples (this one and the tensorflow/tpu example ) use Cloud TPUs, I can reword it to be similar to the suggestion though.

Author: [e2fe072](https://github.com/ray-project/ray/pull/45924/commits/e2fe0720674aca7f7ce0ba09c0eaeb5a510c7168)

## Step 1: Create a Kubernetes cluster with TPUs

Follow [Creating a GKE Cluster with TPUs for KubeRay](kuberay-gke-tpu-cluster-setup) to create a GKE cluster with 1 CPU node and 1 TPU node.

## Step 2: Install the KubeRay operator

> Skip this step if the [Ray Operator Addon](https://cloud.google.com/kubernetes-engine/docs/add-on/ray-on-gke/concepts/overview) is enabled in your GKE cluster. Follow [this document](kuberay-operator-deploy) to install the latest stable KubeRay operator via Helm repository. Multi-host TPU support is provided in KubeRay v1.1.0+. Note that the YAML file in this example uses `serveConfigV2`, which is supported starting from KubeRay v0.6.0.

Reviewer 3: ```suggestion
Skip this step if the [Ray Operator Addon](https://cloud.google.com/kubernetes-engine/docs/add-on/ray-on-gke/concepts/overview) is enabled in your GKE cluster. Follow [Deploy a KubeRay operator](kuberay-operator-deploy) instructions to install the latest stable KubeRay operator from the Helm repository. Multi-host TPU support is available in KubeRay v1.1.0+. Note that the YAML file in this example uses `serveConfigV2`, which KubeRay supports starting from v0.6.0.
```

> ## Step 3: Install the RayService

Reviewer: Maybe use custom resource or `CR` instead? CRD and CR are different.

> ## Step 3: Install the RayService

Author: [48c298a](https://github.com/ray-project/ray/pull/45924/commits/48c298aa8b9f2aa6dce58ad7e9ea7a15163d6a3c)

> ## Step 3: Install the RayService

Author: Done in [1e1fe21](https://github.com/ray-project/ray/pull/45924/commits/1e1fe2146656ae3f04c51e13bb0a203eb5ac4946)

```sh
# Creates a RayCluster with a single-host v4 TPU worker group of 2x2x1 topology.
kubectl apply -f https://raw.githubusercontent.com/ray-project/kuberay/master/ray-operator/config/samples/ray-service.tpu-single-host.yaml
```

KubeRay operator v1.1.0 adds a new `NumOfHosts` field to the RayCluster CR, supporting multi-host worker groups. This field specifies the number of workers to create per replica, with each replica representing a multi-host Pod slice. The value for `NumOfHosts` should match the number of TPU VM hosts that the given `cloud.google.com/gke-tpu-topology` node selector expects. For this example, the Stable Diffusion model is small enough to run on a single TPU host, so `numOfHosts` is set to 1 in the RayService manifest.

> ## Step 4: View the Serve deployment in the Ray dashboard

Reviewer 3: ```suggestion
## Step 4: View the Serve deployment in the Ray Dashboard
```

> Verify the RayService CR is deployed and running

Reviewer 3: ```suggestion
Verify that you deployed the RayService CR and it's running:
```

> Verify the RayService is deployed and running

Reviewer: Use CR or custom resource instead

> Verify the RayService is deployed and running

Author: [1e1fe21](https://github.com/ray-project/ray/pull/45924/commits/1e1fe2146656ae3f04c51e13bb0a203eb5ac4946)

```sh
kubectl get rayservice

# NAME               SERVICE STATUS   NUM SERVE ENDPOINTS
# stable-diffusion-tpu-serve-svc   Running          2
```

> Port-forward the Ray dashboard from the Ray head service. To view the dashboard, open http://localhost:8265/ on your local machine.

Reviewer 3: ```suggestion
Port-forward the Ray Dashboard from the Ray head service. To view the dashboard, open http://localhost:8265/ on your local machine.
```

```sh
kubectl port-forward svc/stable-diffusion-tpu-head-svc 8265:8265 &
```

> Monitor the status of the RayService CR in the Ray dashboard from the the 'Serve' tab. The installed RayService CR should create a running app with the name 'stable_diffusion'. The app should have two deployments, the API ingress which receives input prompts and the Stable Diffusion model server.

Reviewer 3: ```suggestion
Monitor the status of the RayService CR in the Ray Dashboard from the the 'Serve' tab. The installed RayService CR should create a running app with the name 'stable_diffusion'. The app should have two deployments, the API ingress, which receives input prompts, and the Stable Diffusion model server.
```

![serve_dashboard](../images/serve_dashboard.png)

## Step 5: Send text-to-image prompts to the model server

Port forward the Ray Serve service:
```sh
kubectl port-forward svc/stable-diffusion-tpu-serve-svc 8000
```

In a separate terminal, download the Python prompt script:

```sh
curl -LO https://raw.githubusercontent.com/ray-project/serve_config_examples/master/stable_diffusion/stable_diffusion_tpu_req.py
```

Install the required dependencies to run the Python script locally:

```sh
# Create a Python virtual environment.
python3 -m venv myenv
source myenv/bin/activate

pip install numpy pillow requests tqdm
```

Submit a text-to-image prompt to the Stable Diffusion model server:
```sh
python stable_diffusion_tpu_req.py  --save_pictures
```

> * The results of the stable diffusion inference are saved to a file named diffusion_results.png.

Reviewer 3: ```suggestion
* Ray saves the results of the Stable Diffusion inference to a file named diffusion_results.png.
```

Reviewer 3: Please check for technical accuracy, that is, is it Ray that saves the results?

Author: The inference results are returned from the Stable Diffusion server running on Ray, and saved by the prompt script. I'll reword the sentence to "The Python prompt script saves the results..."

Author: [e2fe072](https://github.com/ray-project/ray/pull/45924/commits/e2fe0720674aca7f7ce0ba09c0eaeb5a510c7168)

(kuberay-tpu-stable-diffusion-example)=

# Serve a Stable Diffusion model on GKE with TPUs

> **Note:** The Python files for the Ray Serve application and its client are in the [ray-project/serve_config_examples](https://github.com/ray-project/serve_config_examples). This example is adapted from the [tensorflow/tpu](https://github.com/tensorflow/tpu/tree/master/tools/ray_tpu/src/serve) Cloud TPU example.

## Step 1: Create a Kubernetes cluster with TPUs

Follow [Creating a GKE Cluster with TPUs for KubeRay](kuberay-gke-tpu-cluster-setup) to create a GKE Autopilot or Standard cluster with 1 CPU node and 1 TPU node.

## Step 2: Install the KubeRay operator

> Follow [this document](kuberay-operator-deploy) to install the latest stable KubeRay operator via Helm repository. Multi-host TPU support is provided in KubeRay v1.1.0+. Please note that the YAML file in this example uses `serveConfigV2`, which is supported starting from KubeRay v0.6.0.

Reviewer: It only needs to be installed if the add-on is not enabled.

Author: Done in [2a6a93f](https://github.com/ray-project/ray/pull/45924/commits/2a6a93fef8436a802b85f334d779f1b54c4d2d6a).
